danceradvent/public/articles/2015/7-medium-scale-part-5.pod

```diff
@@ -0,0 +1,208 @@
+thus may never consider whether it is a cohesive design.
+
+This practice of defining Dancer route handlers as one-liners solves
+that. It ensures that every time you adjust the route handlers for your
+site, you're doing so while looking at the neighboring routes; you will
+be defining new routes in context, not independently. This counteracts
+the entropic forces that result in API design drift because it
+implicitly encourages you to define new routes that fit logically into
+the scheme you defined previously. Without this extra Perl function call
+layer, the route definitions are visually broken up, so that you cannot
+see the whole design on a single screen. The resulting URL scheme can
+turn into a mess, as developers hack on the app over the course of
+years.
+
+This is also why I have carefully lined up the blocks and
+L<fat commas|https://en.wikipedia.org/wiki/Fat_comma> in the route
+definitions: I want patterns in the URL scheme to jump out at me.
+Left-justifying everything obscures these patterns.
+
+Now you may be asking, why does all this matter?
+
+The reason is simple: When you make patterns in the URL scheme
+graphically apparent, you may discover some previously hidden practical
+value that was just sitting there waiting to be exploited. It works for
+the same reason that seeing a plot for a numeric data set for the first
+time sometimes creates a clear call to action.
+
+It turns out that the example route scheme we've been defining
+throughout this article series includes a web API; it's just been
+sitting there, waiting to be discovered. With a bit of polishing, it
+might even turn into a proper
+L<REST|https://en.wikipedia.org/wiki/Representational_state_transfer>
+API. Let's try.
+
+Look back up at that route definition block above. What does it remind
+you of? If you said
+L<CRUD|https://en.wikipedia.org/wiki/Create,_read,_update_and_delete>,
+go to the head of the class.
+
+Now, HTTP verbs do not correspond directly to CRUD because HTTP was not
+designed to be a CRUD API. There are competing design philosophies when
+it comes to mapping HTTP verbs to the CRUD model of persistent data
+storage.
+
+I prefer this mapping:
+
+   DB term    | HTTP verb
+ +------------+----------
+ | C = create | POST
+ | R = read   | GET
+ | U = update | PUT
+ | D = delete | DELETE
+
+The R = C<GET> and D = C<DELETE> rules are obvious, but it may not be
+clear to you why the other two rules are correct.
+
+L<The HTTP spec|http://www.w3.org/Protocols/rfc2616/rfc2616.html>
+defines C<POST> and C<PUT> so that they're nearly interchangeable, which
+is how we ended up with competing REST/CRUD API design philosophies.
+
+Some web API designers like to use C<POST> for both create and update
+operations since you can distinguish the cases based on whether the
+HTTP request contains a database row ID: if so, it's a request to update
+an existing record, else it's a record creation request.
```

Comment:
I can't see what that would add to the article, other than confusion. My purpose in the article series is to recommend a single set of cohesive ideas, not to explain all of the alternative ideas, too.

I started to write something about this anyway, as reading through the RFC proved somewhat interesting, but then I realized that, having never used it in production, that I could not justify its use short of simply recommending that people read the RFC. That's no basis for making recommendations.

If you want me to write about `PATCH`, you're going to have to give me a hard use case: an example I can run here where `PUT` does the wrong thing, and changing the code to use `PATCH` fixes it.